hazelcast/src/main/java/com/hazelcast/transaction/impl/TransactionImpl.java

```diff
@@ -190,37 +189,59 @@ public void remove(Object key) {
     }
 
     private void checkThread() {
-        if (checkThreadAccess && threadId != null && threadId.longValue() != Thread.currentThread().getId()) {
+        if (!checkThreadAccess) {
+            return;
+        }
+        if (threadId != null && threadId.longValue() != Thread.currentThread().getId()) {
             throw new IllegalStateException("Transaction cannot span multiple threads!");
         }
+        if (state == ACTIVE && ACTIVE_THREAD_TRANSACTION.get() != this) {
+            throw new IllegalStateException("Transaction is not active on current thread!");
+        }
+    }
+
+    private void checkActiveThreadBoundTransaction() {
+        if (hasThreadBoundTransactions()) {
+            throw new IllegalStateException("Current thread already bound with transaction!");
+        }
     }
 
     @Override
     public void begin() throws IllegalStateException {
         if (state == ACTIVE) {
             throw new IllegalStateException("Transaction is already active");
         }
-        if (TRANSACTION_EXISTS.get() != null) {
-            throw new IllegalStateException("Nested transactions are not allowed!");
-        }
+        checkActiveThreadBoundTransaction();
         startTime = currentTimeMillis();
         backupAddresses = transactionManagerService.pickBackupLogAddresses(durability);
 
         //init caller thread
         if (threadId == null) {
             threadId = Thread.currentThread().getId();
-            setThreadFlag(TRUE);
+            bindThread();
         }
         state = ACTIVE;
         transactionManagerService.startCount.inc();
     }
 
-    private void setThreadFlag(Boolean flag) {
+    private boolean hasThreadBoundTransactions() {
+        return ACTIVE_THREAD_TRANSACTION.get() != null;
+    }
+
+    private void setThreadFlag(TransactionImpl transaction) {
         if (checkThreadAccess) {
-            TRANSACTION_EXISTS.set(flag);
+            ACTIVE_THREAD_TRANSACTION.set(transaction);
         }
     }
 
+    private void bindThread() {
+        setThreadFlag(this);
+    }
+
+    private void clearThread() {
+        setThreadFlag(null);
```

Comment:
Do we need to set the `state` here to `NO_TXN`? And back to `ACTIVE` in `bindThread`?

Reply:
this probably will be good, thanks

fixed

Reply:
state is now changing within `SuspendedTransactionImpl.suspend` and `SuspendedTransactionImpl.resume`